### ISSUE TITLE
Cast Exception on trying to set DataType on a DataTemplete

### ISSUE THREAD

**Issue by JAAYapps** [AvaloniaUI/Avalonia#17592]
### Describe the bug

I'm getting a System.InvalidCastException when I am trying to create a projects view for my app.

### To Reproduce

Using this code in a usercontrol.

```
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             xmlns:vm="using:UndertaleModToolAvalonia.ViewModels.StartPageViewModels.DataItemViewModels"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="UndertaleModToolAvalonia.Views.StartPageViews.DataItemViews.DataFileView" x:Name="FileView"
             x:DataType="vm:DataFileViewModel">

  <ItemsControl ItemsSource="{Binding Items}">
    <ItemsControl.ItemsPanel>
      <ItemsPanelTemplate>
        <WrapPanel />
      </ItemsPanelTemplate>
    </ItemsControl.ItemsPanel>
    <ItemsControl.ItemTemplate>
      <DataTemplate DataType="vm:DataFileItemViewModel">
        <Button BorderThickness="0" Command="{Binding #FileView.((vm:DataFileViewModel)DataContext).LoadData}" CommandParameter="{Binding Name}">
          <Button.Styles>
            <Style Selector="Button">
              <Setter Property="Background" Value="Transparent"/>
            </Style>
            <Style Selector="Button:pointerover">
              <Setter Property="Opacity" Value="0.5"/>
              <Setter Property="Background" Value="#D5BF8D"/>
            </Style>
          </Button.Styles>
          <Grid Margin="10" MinWidth="160" MinHeight="160" MaxWidth="160" MaxHeight="160">
            <Grid.RowDefinitions>
              <RowDefinition Height="2*"/>
              <RowDefinition Height="*"/>
            </Grid.RowDefinitions>
            <Grid Grid.Row="0">
              <Image x:Name="DataIcon" Margin="0,-20,0,0" Stretch="None" Source="avares://Assets/icon.ico" />
            </Grid>
            <TextBlock x:Name="fileName" Text="{Binding Preview}" Grid.Row="1" 
                       FontSize="18" TextWrapping="Wrap" TextAlignment="Center" Foreground="#FF362121" 
                       HorizontalAlignment="Stretch" VerticalAlignment="Stretch" LineHeight="20"
                       FontWeight="Bold"/>
          </Grid>
        </Button>
      </DataTemplate>
    </ItemsControl.ItemTemplate>
  </ItemsControl>
</UserControl>
```

Then run with debugging. You should get System.InvalidCastException: Unable to cast object of type 'UndertaleModToolAvalonia.ViewModels.StartPageViewModels.DataItemViewModels.DataFileItemViewModel' to type 'UndertaleModToolAvalonia.ViewModels.StartPageViewModels.DataItemViewModels.DataFileViewModel'.
   at CompiledAvaloniaXaml.XamlIlHelpers.UndertaleModToolAvalonia.ViewModels.StartPageViewModels.DataItemViewModels.DataFileViewModel,UndertaleModToolAvalonia.Items!Getter(Object)
   at Avalonia.Data.Core.ClrPropertyInfo.Get(Object target)
   at Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindings.InpcPropertyAccessor.get_Value()
   at Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindings.InpcPropertyAccessor.SendCurrentValue()

### Expected behavior

Properly use the viewmodels I set.

### Avalonia version

11.2.1

### OS

Linux

### Additional context

Not sure why this happens. I was following the To Do List sample to create this.

**Comment by timunie** [comment]
@JAAYapps please make sure your items are of the correct type. If you read the error message carefully, it tells you which type it expects and which the item had.

**Comment by JAAYapps** [comment]
I had to deeply investigate what was going on. Wish I wasn't visually impaired, but I eventually found it. Sorry, I just couldn't see it clearly and went a haunt finding my issue. Thanks for letting me know it was on my end, I was going crazy trying to find my mistake. Turns out that the it was the wrong type in the viewmodel class where I had my ObservableCollection. Wish I had a stacktrace that went through the viewmodel as well, but I guess that's a .net thing.
